// openclaw/dist/plugin-sdk/channel-setup.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../utils-CS0Ikux6.js";
import { t as formatDocsLink } from "../links-8xRhWBQL.js";
import { g as DEFAULT_ACCOUNT_ID } from "../session-key-DAhnzjyr.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../path-alias-guards-CwRM04O1.js";
import "../setup-binary-C2MYZTCc.js";
import "../archive-pOxyr9r4.js";
import "../fs-safe-DZyOgpPN.js";
import "../signal-cli-install-fVOL9FvX.js";
import { C as createTopLevelChannelDmPolicy, rt as setSetupChannelEnabled, st as splitSetupEntries } from "../setup-wizard-proxy-DoOpOF8S.js";
import "../setup-BI3ohdrT.js";
import { n as createOptionalChannelSetupAdapter, r as createOptionalChannelSetupWizard, t as createOptionalChannelSetupSurface } from "../channel-setup-C5JbULY6.js";
export { DEFAULT_ACCOUNT_ID, createOptionalChannelSetupAdapter, createOptionalChannelSetupSurface, createOptionalChannelSetupWizard, createTopLevelChannelDmPolicy, formatDocsLink, setSetupChannelEnabled, splitSetupEntries };
