// openclaw/dist/plugin-sdk/discord-core.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import { n as DiscordConfigSchema } from "../zod-schema.providers-core-COy5nBQ0.js";
import { r as getChatChannelMeta } from "../registry-C5UkPpaO.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import { a as withNormalizedTimestamp } from "../date-time-DcBRR7Wo.js";
import { c as jsonResult, d as readNumberParam, f as readReactionParams, h as readStringParam, l as parseAvailableTags, p as readStringArrayParam } from "../common-zReYuoMM.js";
import "../image-ops-By4YfyWa.js";
import "../path-alias-guards-CwRM04O1.js";
import { t as assertMediaNotDataUrl } from "../sandbox-paths-Cn8i8Jxm.js";
import "../mime-CsUtzcn2.js";
import { r as buildChannelConfigSchema } from "../config-schema-CPzP2Ds6.js";
import "../channel-plugin-common-ClptUZHU.js";
import { r as resolvePollMaxSelections } from "../polls-BSvCQ2Fu.js";
import "../discord-core-DLbIyXWy.js";
export { DiscordConfigSchema, assertMediaNotDataUrl, buildChannelConfigSchema, getChatChannelMeta, jsonResult, parseAvailableTags, readNumberParam, readReactionParams, readStringArrayParam, readStringParam, resolvePollMaxSelections, withNormalizedTimestamp };
