// openclaw/dist/plugin-sdk/infra-runtime.js
import "../redact-BDinS1q9.js";
import { a as hasErrnoCode, i as formatUncaughtError, n as extractErrorCode, o as isErrno, r as formatErrorMessage, s as readErrorName, t as collectErrorGraphCandidates } from "../errors-BxyFnvP3.js";
import "../unhandled-rejections-CDJ8dOVP.js";
import "../logger-kwZIqwuw.js";
import { C as resolveRequiredHomeDir, S as resolveOsHomeRelativePath, b as resolveHomeRelativePath, v as expandHomePrefix, w as resolveRequiredOsHomeDir, x as resolveOsHomeDir, y as resolveEffectiveHomeDir } from "../paths-ViKUYWUK.js";
import { n as resolvePreferredOpenClawTmpDir, t as POSIX_OPENCLAW_TMP_DIR } from "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import { i as normalizeZaiEnv, n as logAcceptedEnvOption, r as normalizeEnv, t as isTruthyEnvValue } from "../env-Dnra1IpT.js";
import "../warning-filter-CBhOcgHd.js";
import "../utils-CS0Ikux6.js";
import "../links-8xRhWBQL.js";
import "../paths-C4Oy7wjk.js";
import "../auth-profiles-DRuJBw5y.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import { $ as splitCommandChain, Et as DEFAULT_SAFE_BINS, J as buildEnforcedShellCommand, K as analyzeArgvCommand, Q as resolvePlannedSegmentArgv, X as buildSafeShellCommand, Y as buildSafeBinsShellCommand, Z as isWindowsPlatform, at as resolveCommandResolution, ct as resolveExecutionTargetResolution, dt as resolvePolicyTargetResolution, et as splitCommandChainWithOperators, it as resolveApprovalAuditCandidatePath, lt as resolvePolicyAllowlistCandidatePath, nt as parseExecArgvToken, ot as resolveCommandResolutionFromArgv, q as analyzeShellCommand, rt as resolveAllowlistCandidatePath, st as resolveExecutionTargetCandidatePath, tt as matchAllowlist, ut as resolvePolicyTargetCandidatePath } from "../io-y3Az_Onx.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import { C as normalizeScpRemotePath, S as normalizeScpRemoteHost, b as isSafeScpRemoteHost, x as isSafeScpRemotePath } from "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import "../config-D4zN4BRl.js";
import { n as resetFileLockStateForTest, r as withFileLock, t as acquireFileLock } from "../file-lock-DEiVq2ow.js";
import "../audit-fs-DTQIjp2r.js";
import "../resolve-B7S4PlnO.js";
import "../profiles-COPO-hHI.js";
import "../tailscale-CxY-91GZ.js";
import "../tailnet-DHiKSG7V.js";
import "../net-JAg9zL0G.js";
import "../auth-Buq0Niri.js";
import "../credentials-D8pfRimu.js";
import "../message-channel-BliByQBl.js";
import "../store-C5UK26Ce.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import "../sessions-D9ZgHCHb.js";
import "../paths-rhN9LKM_.js";
import "../session-write-lock-va3qZE6f.js";
import { i as writeTextAtomic, n as readJsonFile, r as writeJsonAtomic, t as createAsyncLock } from "../json-files-2-D1jkJB.js";
import "../method-scopes-BiEi0X2g.js";
import "../call-CQbSO4Fr.js";
import { a as resetWSLStateForTests, i as isWSLSync, n as isWSL2Sync, r as isWSLEnv, t as isWSL } from "../wsl-BVCo-dpy.js";
import "../prompt-style-C4CV4Y_p.js";
import "../ports-lsof-BLEgL_Gn.js";
import "../restart-stale-pids-BsPvlGJd.js";
import "../ports-FldL_Lhq.js";
import "../logging-CznuqaMu.js";
import "../commands-C6oAw_H-.js";
import "../issue-format-BtLIwOtN.js";
import "../identity-Dp-UIak4.js";
import "../heartbeat-Bdfr_onc.js";
import { $y as resolveExecApprovalCommandDisplay, Ay as PROXY_FETCH_PROXY_URL, By as resolveFetch, Cy as TELEGRAM_RETRY_DEFAULTS, Dy as ensureGlobalUndiciEnvProxyDispatcher, Ey as DEFAULT_UNDICI_STREAM_TIMEOUT_MS, Fy as emitHeartbeatEvent, Hy as resolveExecApprovalSessionTarget, Iy as getLastHeartbeatEvent, Ly as onHeartbeatEvent, My as makeProxyFetch, Ny as resolveProxyFetchFromEnv, Oy as ensureGlobalUndiciStreamTimeouts, Py as resolveHeartbeatVisibility, Qy as getExecApprovalReplyMetadata, Ry as resetHeartbeatEventsForTest, Sy as waitForTransportReady, Ty as createTelegramRetryRunner, Vy as wrapFetchWithAbortSignal, Xy as buildExecApprovalUnavailableReplyPayload, Yy as buildExecApprovalPendingReplyPayload, Zy as getExecApprovalApproverDmNoticeText, ab as recordChannelActivity, eb as sanitizeExecApprovalDisplayText, ib as getChannelActivity, jy as getProxyUrlFromFetch, ky as resetGlobalUndiciStreamTimeoutsForTests, nb as matchesDiagnosticFlag, ob as resetChannelActivityForTest, rb as resolveDiagnosticFlags, tb as isDiagnosticFlagEnabled, wy as createRateLimitRetryRunner, zy as resolveIndicatorType } from "../pi-embedded-CbCYZxIb.js";
import "../internal-hooks-CIUghoK4.js";
import { n as generateSecureUuid, t as generateSecureToken } from "../secure-random-2W4K8IQm.js";
import "../multimodal-B8yEZcl5.js";
import "../memory-search-B_aq08y5.js";
import "../provider-catalog-rVN7ZbR5.js";
import "../secret-input-DXoOLNgb.js";
import "../bindings-CFudGB9Z.js";
import "../resolve-route-DwFWP4JW.js";
import "../routing-zOQ0wwtV.js";
import "../identity-file-DVLib087.js";
import { n as resolveAgentOutboundIdentity, r as resolveOutboundSendDep, t as normalizeOutboundIdentity } from "../outbound-runtime-m-y0nd6F.js";
import "../provider-env-vars-CpUJqT5v.js";
import "../provider-auth-input-Bus9W_qh.js";
import "../provider-model-minimax-DSmX4BNO.js";
import { D as resolveRetryConfig, O as retryAsync } from "../provider-models-BKjzpTsb.js";
import "../anthropic-vertex-provider-B2x4YywE.js";
import "../provider-catalog-BTgF8VyO.js";
import "../provider-catalog-D0IlX18O.js";
import "../provider-catalog-BTmUX0Mp.js";
import "../provider-catalog-sph_Rl63.js";
import "../provider-catalog-C7tUyBwV.js";
import "../provider-catalog-C8dPZpHW.js";
import "../provider-catalog-BJNff1VI.js";
import "../provider-catalog-IAdymoqT.js";
import "../provider-catalog-Bjc7evzh.js";
import "../models-config.providers.discovery-CjkO08q6.js";
import "../text-runtime-DVXt5xhi.js";
import "../tool-catalog-BjSY4C4F.js";
import "../docker-Bhjg8g2t.js";
import "../sandbox-DTlKNieF.js";
import "../common-zReYuoMM.js";
import "../image-ops-By4YfyWa.js";
import "../thinking-CJoSRIGR.js";
import { a as trySafeFileURLToPath, i as safeFileURLToPath, n as basenameFromMediaSource, r as isWindowsNetworkPath, t as assertNoWindowsNetworkPath } from "../local-file-access-D6qxzMHn.js";
import "../path-alias-guards-CwRM04O1.js";
import "../sandbox-paths-Cn8i8Jxm.js";
import "../channel-actions-fV93bLjV.js";
import "../mime-CsUtzcn2.js";
import { i as resolveEnvHttpProxyUrl, n as hasEnvHttpProxyConfigured, r as hasProxyEnvConfigured, t as PROXY_ENV_KEYS } from "../proxy-env-BHEDx2JD.js";
import { a as createPinnedLookup, c as isPrivateIpAddress, d as resolvePinnedHostnameWithPolicy, i as createPinnedDispatcher, l as isPrivateNetworkAllowedByPolicy, n as assertPublicHostname, o as isBlockedHostname, p as normalizeHostname, r as closeDispatcher, s as isBlockedHostnameOrIp, t as SsrFBlockedError, u as resolvePinnedHostname } from "../ssrf-0bPJMoZR.js";
import { i as withTrustedEnvProxyGuardedFetchMode, n as fetchWithSsrFGuard, r as withStrictGuardedFetchMode, t as GUARDED_FETCH_MODE } from "../fetch-guard-CYl1q2XH.js";
import "../provider-web-search-COZNygbd.js";
import "../manager-Bw8JrihM.js";
import "../config-schema-CPzP2Ds6.js";
import "../channel-plugin-common-ClptUZHU.js";
import "../status-helpers-Cflmo90d.js";
import "../conversation-runtime-nmWgJSXS.js";
import "../runtime-whatsapp-boundary-XjdkYMz_.js";
import "../pairing-store-BGKLrjo6.js";
import "../json-store--JDX3zNL.js";
import { i as tryReadSecretFileSync, n as loadSecretFileSync, r as readSecretFileSync, t as DEFAULT_SECRET_FILE_MAX_BYTES } from "../secret-file-Bif7RlZ-.js";
import "../channel-config-schema-CgpS17zB.js";
import "../setup-binary-C2MYZTCc.js";
import "../archive-pOxyr9r4.js";
import { a as openFileWithinRoot, c as readLocalFileSafely, d as writeFileFromPathWithinRoot, f as writeFileWithinRoot, i as createRootScopedReadFile, l as readPathWithinRoot, n as appendFileWithinRoot, o as openWritableFileWithinRoot, r as copyFileWithinRoot, s as readFileWithinRoot, t as SafeOpenError, u as resolveOpenedFileRealPathForHandle } from "../fs-safe-DZyOgpPN.js";
import "../signal-cli-install-fVOL9FvX.js";
import "../setup-wizard-proxy-DoOpOF8S.js";
import "../setup-BI3ohdrT.js";
import "../dm-policy-shared-DtcQProI.js";
import "../runtime-env-DQOYrXVl.js";
import "../temp-path-B91lt0LF.js";
import { i as resetDiagnosticEventsForTest, n as isDiagnosticsEnabled, r as onDiagnosticEvent, t as emitDiagnosticEvent } from "../diagnostic-events-DLiZiw1x.js";
import "../diagnostic-CDTtMQBT.js";
import "../logging-CCTKg_fh.js";
import "../config-runtime-BYNizC50.js";
import { f as computeBackoff, p as sleepWithAbort } from "../level-overrides-DzIzQ3ZY.js";
import { t as pruneMapToMaxSize } from "../map-size-BQ5KZ2st.js";
import { n as resolveGlobalDedupeCache, t as createDedupeCache } from "../dedupe-DML1ACVm.js";
import { C as isSafeBinUsage, E as resolveSafeBins, S as evaluateShellAllowlist, T as resolveAllowAlwaysPatterns, _ as resolveExecApprovalsFromFile, a as maxAsk, b as saveExecApprovals, c as normalizeExecApprovals, d as normalizeExecSecurity, f as readExecApprovalsSnapshot, g as resolveExecApprovals, h as requiresExecApproval, i as loadExecApprovals, l as normalizeExecAsk, m as requestExecApprovalViaSocket, n as addAllowlistEntry, o as mergeExecApprovalsSocketDefaults, p as recordAllowlistUse, r as ensureExecApprovals, s as minSecurity, t as DEFAULT_EXEC_APPROVAL_TIMEOUT_MS, u as normalizeExecHost, v as resolveExecApprovalsPath, w as normalizeSafeBins, x as evaluateExecAllowlist, y as resolveExecApprovalsSocketPath } from "../exec-approvals-BF_Qfdq8.js";
import { i as formatDurationSeconds, n as formatDurationHuman, r as formatDurationPrecise, t as formatDurationCompact } from "../format-duration-DwyQDycT.js";
import { A as DEFAULT_WEBHOOK_MAX_BODY_BYTES, F as readRequestBodyWithLimit, I as requestBodyErrorToText, M as installRequestBodyLimitGuard, N as isRequestBodyLimitError, P as readJsonBodyWithLimit, j as RequestBodyLimitError, k as DEFAULT_WEBHOOK_BODY_TIMEOUT_MS } from "../webhook-ingress-D-H5frou.js";
import { a as isSystemEventContextChanged, c as resetSystemEventsForTest, i as hasSystemEvents, n as drainSystemEvents, o as peekSystemEventEntries, r as enqueueSystemEvent, s as peekSystemEvents, t as drainSystemEventEntries } from "../system-events-BU_NYynX.js";
import { n as hasSystemMark, r as prefixSystemMessage, t as SYSTEM_MARK } from "../system-message-C205nJsr.js";
import { n as createRuntimeOutboundDelegates } from "../runtime-forwarders-Cv9Bdwov.js";
import { a as ssrfPolicyFromAllowPrivateNetwork, i as normalizeHostnameSuffixAllowlist, n as buildHostnameAllowlistPolicyFromSuffixAllowlist, r as isHttpsUrlAllowedByHostnameSuffixAllowlist, t as assertHttpUrlTargetsPrivateNetwork } from "../ssrf-policy-DYINyIKC.js";
import "../hook-runtime-BUYmDbaA.js";
import "../templating-BDXRiJUh.js";
import "../channel-reply-pipeline-BxH4-AP2.js";
import "../reply-history-n7nOPLta.js";
import "../commands-registry.data-BFXJWoqi.js";
import "../commands-registry-Bh9pJ817.js";
import "../frontmatter-CP5371Rm.js";
import "../env-overrides-BNSo89u5.js";
import "../skills-M0AZJeXx.js";
import "../skills-remote-BHxDA4Jw.js";
import "../workspace-dirs-DqDLiRvZ.js";
import "../pairing-token-DAszGmwD.js";
import "../skill-commands-D73dGR_l.js";
import "../config-CLRQPUSt.js";
import "../routes-XT2JfvT-.js";
import "../ssh-tunnel-LM52qY6N.js";
import "../server-middleware-BP59C6FL.js";
import "../provider-auth-ref-CQEp9sp3.js";
import "../provider-auth-helpers-Do3RFvdi.js";
import "../provider-api-key-auth-BEeRW3dq.js";
import "../pairing-labels-D27Y4Vv6.js";
import "../directory-runtime-D9Y42mW-.js";
import "../read-only-account-inspect-CJPRBFj7.js";
import "../src-BoNU6eQs.js";
import "../web-media-PKy3RbRG.js";
import "../media-understanding-BDOSvuM9.js";
import "../web-media-CiSbdGPs.js";
import "../state-paths-Dvb3Mfl3.js";
import "../llm-task-Br6Ixjlh.js";
import "../pi-model-discovery-DO3tFksS.js";
import "../exec-inline-eval-CpmxWJsW.js";
import "../target-registry-9OUOvlxj.js";
import "../external-content-DZI9vlJn.js";
import "../brave-onFD06fq.js";
import "../duckduckgo-CZuAOfiQ.js";
import "../exa-5bwgY04k.js";
import "../security-runtime-xlw0Fxx2.js";
import "../provider-usage-ClxTGn5c.js";
import "../provider-models-Bjv0V4OP.js";
import "../media-understanding-provider-DllGbg-v.js";
import "../media-understanding-provider-BDq4rFL-.js";
import "../provider-onboard-CSnPvkMM.js";
import "../onboard-BU8aRvq7.js";
import "../perplexity-web-search-provider-CI38RNUU.js";
import "../perplexity-CMZbsreI.js";
import "../model-definitions-CLagLj0g.js";
import "../onboard-0P7GBI2D.js";
import "../provider-catalog-CKHo9CRM.js";
import "../provider-models-Dmjtw6kn.js";
import "../stream-C-KGzzT0.js";
import "../web-search-CO1hvu_l.js";
import "../stagger-CFqdtCYL.js";
import "../command-secret-targets-DLs8jC2H.js";
import "../delivery-queue-DHURMSqX.js";
import "../channel-summary-Ctrv7moa.js";
import "../follow-redirects-CVAaqIk4.js";
import "../tool-policy-match-DQWWRSN4.js";
import "../runtime-CAdCdG-M.js";
import "../channel-status-BfM90KqH.js";
import "../discord-core-DLbIyXWy.js";
import "../cli-runtime-Cx1v7N7v.js";
import "../config-presence-DA3j8Nmi.js";
import "../query-expansion-BOJBTXsp.js";
import "../search-manager-DNBHwaee.js";
import "../acp-runtime-BHZj-SLU.js";
import "../telegram-core-DD8gCuaX.js";
import "../audit-BI48Fi3j.js";
import "../gateway-runtime-Dn2qAGY7.js";
import "../connection-auth-Nl2a3tDb.js";
import "../mcp-config-CDIWSqVk.js";
export { DEFAULT_EXEC_APPROVAL_TIMEOUT_MS, DEFAULT_SAFE_BINS, DEFAULT_SECRET_FILE_MAX_BYTES, DEFAULT_UNDICI_STREAM_TIMEOUT_MS, DEFAULT_WEBHOOK_BODY_TIMEOUT_MS, DEFAULT_WEBHOOK_MAX_BODY_BYTES, GUARDED_FETCH_MODE, POSIX_OPENCLAW_TMP_DIR, PROXY_ENV_KEYS, PROXY_FETCH_PROXY_URL, RequestBodyLimitError, SYSTEM_MARK, SafeOpenError, SsrFBlockedError, TELEGRAM_RETRY_DEFAULTS, acquireFileLock, addAllowlistEntry, analyzeArgvCommand, analyzeShellCommand, appendFileWithinRoot, assertHttpUrlTargetsPrivateNetwork, assertNoWindowsNetworkPath, assertPublicHostname, basenameFromMediaSource, buildEnforcedShellCommand, buildExecApprovalPendingReplyPayload, buildExecApprovalUnavailableReplyPayload, buildHostnameAllowlistPolicyFromSuffixAllowlist, buildSafeBinsShellCommand, buildSafeShellCommand, closeDispatcher, collectErrorGraphCandidates, computeBackoff, copyFileWithinRoot, createAsyncLock, createDedupeCache, createPinnedDispatcher, createPinnedLookup, createRateLimitRetryRunner, createRootScopedReadFile, createRuntimeOutboundDelegates, createTelegramRetryRunner, drainSystemEventEntries, drainSystemEvents, emitDiagnosticEvent, emitHeartbeatEvent, enqueueSystemEvent, ensureExecApprovals, ensureGlobalUndiciEnvProxyDispatcher, ensureGlobalUndiciStreamTimeouts, evaluateExecAllowlist, evaluateShellAllowlist, expandHomePrefix, extractErrorCode, fetchWithSsrFGuard, formatDurationCompact, formatDurationHuman, formatDurationPrecise, formatDurationSeconds, formatErrorMessage, formatUncaughtError, generateSecureToken, generateSecureUuid, getChannelActivity, getExecApprovalApproverDmNoticeText, getExecApprovalReplyMetadata, getLastHeartbeatEvent, getProxyUrlFromFetch, hasEnvHttpProxyConfigured, hasErrnoCode, hasProxyEnvConfigured, hasSystemEvents, hasSystemMark, installRequestBodyLimitGuard, isBlockedHostname, isBlockedHostnameOrIp, isDiagnosticFlagEnabled, isDiagnosticsEnabled, isErrno, isHttpsUrlAllowedByHostnameSuffixAllowlist, isPrivateIpAddress, isPrivateNetworkAllowedByPolicy, isRequestBodyLimitError, isSafeBinUsage, isSafeScpRemoteHost, isSafeScpRemotePath, isSystemEventContextChanged, isTruthyEnvValue, isWSL, isWSL2Sync, isWSLEnv, isWSLSync, isWindowsNetworkPath, isWindowsPlatform, loadExecApprovals, loadSecretFileSync, logAcceptedEnvOption, makeProxyFetch, matchAllowlist, matchesDiagnosticFlag, maxAsk, mergeExecApprovalsSocketDefaults, minSecurity, normalizeEnv, normalizeExecApprovals, normalizeExecAsk, normalizeExecHost, normalizeExecSecurity, normalizeHostname, normalizeHostnameSuffixAllowlist, normalizeOutboundIdentity, normalizeSafeBins, normalizeScpRemoteHost, normalizeScpRemotePath, normalizeZaiEnv, onDiagnosticEvent, onHeartbeatEvent, openFileWithinRoot, openWritableFileWithinRoot, parseExecArgvToken, peekSystemEventEntries, peekSystemEvents, prefixSystemMessage, pruneMapToMaxSize, readErrorName, readExecApprovalsSnapshot, readFileWithinRoot, readJsonBodyWithLimit, readJsonFile, readLocalFileSafely, readPathWithinRoot, readRequestBodyWithLimit, readSecretFileSync, recordAllowlistUse, recordChannelActivity, requestBodyErrorToText, requestExecApprovalViaSocket, requiresExecApproval, resetChannelActivityForTest, resetDiagnosticEventsForTest, resetFileLockStateForTest, resetGlobalUndiciStreamTimeoutsForTests, resetHeartbeatEventsForTest, resetSystemEventsForTest, resetWSLStateForTests, resolveAgentOutboundIdentity, resolveAllowAlwaysPatterns, resolveAllowlistCandidatePath, resolveApprovalAuditCandidatePath, resolveCommandResolution, resolveCommandResolutionFromArgv, resolveDiagnosticFlags, resolveEffectiveHomeDir, resolveEnvHttpProxyUrl, resolveExecApprovalCommandDisplay, resolveExecApprovalSessionTarget, resolveExecApprovals, resolveExecApprovalsFromFile, resolveExecApprovalsPath, resolveExecApprovalsSocketPath, resolveExecutionTargetCandidatePath, resolveExecutionTargetResolution, resolveFetch, resolveGlobalDedupeCache, resolveHeartbeatVisibility, resolveHomeRelativePath, resolveIndicatorType, resolveOpenedFileRealPathForHandle, resolveOsHomeDir, resolveOsHomeRelativePath, resolveOutboundSendDep, resolvePinnedHostname, resolvePinnedHostnameWithPolicy, resolvePlannedSegmentArgv, resolvePolicyAllowlistCandidatePath, resolvePolicyTargetCandidatePath, resolvePolicyTargetResolution, resolvePreferredOpenClawTmpDir, resolveProxyFetchFromEnv, resolveRequiredHomeDir, resolveRequiredOsHomeDir, resolveRetryConfig, resolveSafeBins, retryAsync, safeFileURLToPath, sanitizeExecApprovalDisplayText, saveExecApprovals, sleepWithAbort, splitCommandChain, splitCommandChainWithOperators, ssrfPolicyFromAllowPrivateNetwork, tryReadSecretFileSync, trySafeFileURLToPath, waitForTransportReady, withFileLock, withStrictGuardedFetchMode, withTrustedEnvProxyGuardedFetchMode, wrapFetchWithAbortSignal, writeFileFromPathWithinRoot, writeFileWithinRoot, writeJsonAtomic, writeTextAtomic };
