// openclaw/dist/plugin-sdk/media-understanding.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../logger-BmpSCz93.js";
import "../ip-Ce8EDTBZ.js";
import "../ssrf-0bPJMoZR.js";
import "../fetch-guard-CYl1q2XH.js";
import { c as describeImagesWithModel, d as postJsonRequest, f as postTranscriptionRequest, l as assertOkOrThrowHttpError, n as deepgramMediaUnderstandingProvider, o as transcribeOpenAiCompatibleAudio, p as requireTranscriptionText, s as describeImageWithModel, t as groqMediaUnderstandingProvider, u as normalizeBaseUrl } from "../media-understanding-BDOSvuM9.js";
export { assertOkOrThrowHttpError, deepgramMediaUnderstandingProvider, describeImageWithModel, describeImagesWithModel, groqMediaUnderstandingProvider, normalizeBaseUrl, postJsonRequest, postTranscriptionRequest, requireTranscriptionText, transcribeOpenAiCompatibleAudio };
