// openclaw/dist/extensions/modelstudio/onboard.js
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../utils-CS0Ikux6.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../provider-model-minimax-DSmX4BNO.js";
import "../../provider-models-BKjzpTsb.js";
import "../../anthropic-vertex-provider-B2x4YywE.js";
import "../../provider-onboard-CSnPvkMM.js";
import { a as MODELSTUDIO_GLOBAL_BASE_URL, i as MODELSTUDIO_DEFAULT_MODEL_REF, o as MODELSTUDIO_STANDARD_CN_BASE_URL, s as MODELSTUDIO_STANDARD_GLOBAL_BASE_URL, t as MODELSTUDIO_CN_BASE_URL } from "../../model-definitions-CkLDjdog.js";
import { a as applyModelStudioStandardConfig, c as applyModelStudioStandardProviderConfigCn, i as applyModelStudioProviderConfigCn, n as applyModelStudioConfigCn, o as applyModelStudioStandardConfigCn, r as applyModelStudioProviderConfig, s as applyModelStudioStandardProviderConfig, t as applyModelStudioConfig } from "../../onboard-dr3al9Ns.js";
export { MODELSTUDIO_CN_BASE_URL, MODELSTUDIO_DEFAULT_MODEL_REF, MODELSTUDIO_GLOBAL_BASE_URL, MODELSTUDIO_STANDARD_CN_BASE_URL, MODELSTUDIO_STANDARD_GLOBAL_BASE_URL, applyModelStudioConfig, applyModelStudioConfigCn, applyModelStudioProviderConfig, applyModelStudioProviderConfigCn, applyModelStudioStandardConfig, applyModelStudioStandardConfigCn, applyModelStudioStandardProviderConfig, applyModelStudioStandardProviderConfigCn };
