// openclaw/dist/extensions/kilocode/shared.js
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../utils-CS0Ikux6.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import { C as KILOCODE_MODEL_CATALOG, S as KILOCODE_DEFAULT_MODEL_REF, _ as KILOCODE_DEFAULT_CONTEXT_WINDOW, b as KILOCODE_DEFAULT_MODEL_ID, g as KILOCODE_BASE_URL, v as KILOCODE_DEFAULT_COST, x as KILOCODE_DEFAULT_MODEL_NAME, y as KILOCODE_DEFAULT_MAX_TOKENS } from "../../provider-model-minimax-DSmX4BNO.js";
import "../../provider-models-BKjzpTsb.js";
import "../../anthropic-vertex-provider-B2x4YywE.js";
export { KILOCODE_BASE_URL, KILOCODE_DEFAULT_CONTEXT_WINDOW, KILOCODE_DEFAULT_COST, KILOCODE_DEFAULT_MAX_TOKENS, KILOCODE_DEFAULT_MODEL_ID, KILOCODE_DEFAULT_MODEL_NAME, KILOCODE_DEFAULT_MODEL_REF, KILOCODE_MODEL_CATALOG };
