// openclaw/dist/extensions/minimax/model-definitions.js
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../utils-CS0Ikux6.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../provider-model-minimax-DSmX4BNO.js";
import "../../provider-models-BKjzpTsb.js";
import "../../anthropic-vertex-provider-B2x4YywE.js";
import { a as MINIMAX_API_COST, c as MINIMAX_HOSTED_MODEL_ID, d as buildMinimaxApiModelDefinition, f as buildMinimaxModelDefinition, i as MINIMAX_API_BASE_URL, l as MINIMAX_HOSTED_MODEL_REF, n as DEFAULT_MINIMAX_CONTEXT_WINDOW, o as MINIMAX_CN_API_BASE_URL, r as DEFAULT_MINIMAX_MAX_TOKENS, s as MINIMAX_HOSTED_COST, t as DEFAULT_MINIMAX_BASE_URL, u as MINIMAX_LM_STUDIO_COST } from "../../model-definitions-ih_Xt2c8.js";
export { DEFAULT_MINIMAX_BASE_URL, DEFAULT_MINIMAX_CONTEXT_WINDOW, DEFAULT_MINIMAX_MAX_TOKENS, MINIMAX_API_BASE_URL, MINIMAX_API_COST, MINIMAX_CN_API_BASE_URL, MINIMAX_HOSTED_COST, MINIMAX_HOSTED_MODEL_ID, MINIMAX_HOSTED_MODEL_REF, MINIMAX_LM_STUDIO_COST, buildMinimaxApiModelDefinition, buildMinimaxModelDefinition };
