// openclaw/dist/extensions/anthropic/index.js
import "../../redact-BDinS1q9.js";
import "../../errors-BxyFnvP3.js";
import "../../unhandled-rejections-CDJ8dOVP.js";
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../boolean-C3GkJetE.js";
import "../../env-Dnra1IpT.js";
import "../../warning-filter-CBhOcgHd.js";
import "../../utils-CS0Ikux6.js";
import "../../links-8xRhWBQL.js";
import { s as CLAUDE_CLI_PROFILE_ID } from "../../paths-C4Oy7wjk.js";
import { g as suggestOAuthProfileIdForLegacyDefault } from "../../auth-profiles-DRuJBw5y.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../io-y3Az_Onx.js";
import "../../shell-env-BOu7XeT_.js";
import "../../safe-text-yavot2qw.js";
import "../../version-CD3oP1-d.js";
import "../../env-substitution-O6uabUmO.js";
import "../../includes-Bj3eLUWH.js";
import "../../zod-schema.providers-core-COy5nBQ0.js";
import "../../legacy-web-search-BB-ZHEhz.js";
import "../../registry-C5UkPpaO.js";
import "../../config-state-Br0ucqMb.js";
import "../../min-host-version-RMBWtIAR.js";
import "../../manifest-registry-B5JNQdOM.js";
import "../../runtime-guard-PhQ6PwQa.js";
import "../../avatar-policy-B5nOfso_.js";
import "../../ip-Ce8EDTBZ.js";
import { t as parseDurationMs } from "../../parse-duration-D6oKXEgx.js";
import "../../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../../zod-schema.core-BuVz8Rk7.js";
import "../../config-D4zN4BRl.js";
import "../../file-lock-DEiVq2ow.js";
import "../../audit-fs-DTQIjp2r.js";
import "../../resolve-B7S4PlnO.js";
import { n as normalizeSecretInput } from "../../normalize-secret-input-BsX1agd_.js";
import { a as upsertAuthProfile, n as listProfilesForProvider } from "../../profiles-COPO-hHI.js";
import { t as formatCliCommand } from "../../command-format-DBQUhADm.js";
import "../../tailscale-CxY-91GZ.js";
import "../../tailnet-DHiKSG7V.js";
import "../../net-JAg9zL0G.js";
import "../../auth-Buq0Niri.js";
import "../../credentials-D8pfRimu.js";
import "../../message-channel-BliByQBl.js";
import "../../store-C5UK26Ce.js";
import "../../runtime-Iz8uZ7EU.js";
import "../../plugins-B09-vgme.js";
import "../../sessions-D9ZgHCHb.js";
import "../../paths-rhN9LKM_.js";
import "../../session-write-lock-va3qZE6f.js";
import "../../method-scopes-BiEi0X2g.js";
import "../../call-CQbSO4Fr.js";
import "../../prompt-style-C4CV4Y_p.js";
import "../../ports-lsof-BLEgL_Gn.js";
import "../../restart-stale-pids-BsPvlGJd.js";
import "../../ports-FldL_Lhq.js";
import "../../logging-CznuqaMu.js";
import "../../commands-C6oAw_H-.js";
import "../../issue-format-BtLIwOtN.js";
import "../../identity-Dp-UIak4.js";
import "../../heartbeat-Bdfr_onc.js";
import { nv as buildTokenProfileId, rv as validateAnthropicSetupToken } from "../../pi-embedded-CbCYZxIb.js";
import "../../internal-hooks-CIUghoK4.js";
import "../../multimodal-B8yEZcl5.js";
import "../../memory-search-B_aq08y5.js";
import "../../provider-catalog-rVN7ZbR5.js";
import "../../secret-input-DXoOLNgb.js";
import "../../bindings-CFudGB9Z.js";
import "../../resolve-route-DwFWP4JW.js";
import "../../routing-zOQ0wwtV.js";
import "../../identity-file-DVLib087.js";
import "../../outbound-runtime-m-y0nd6F.js";
import "../../provider-env-vars-CpUJqT5v.js";
import { a as normalizeSecretInputModeInput } from "../../provider-auth-input-Bus9W_qh.js";
import { p as normalizeModelCompat } from "../../provider-model-minimax-DSmX4BNO.js";
import "../../provider-models-BKjzpTsb.js";
import "../../anthropic-vertex-provider-B2x4YywE.js";
import "../../provider-catalog-BTgF8VyO.js";
import "../../provider-catalog-D0IlX18O.js";
import "../../provider-catalog-BTmUX0Mp.js";
import "../../provider-catalog-sph_Rl63.js";
import "../../provider-catalog-C7tUyBwV.js";
import "../../provider-catalog-C8dPZpHW.js";
import "../../provider-catalog-BJNff1VI.js";
import "../../provider-catalog-IAdymoqT.js";
import "../../provider-catalog-Bjc7evzh.js";
import "../../models-config.providers.discovery-CjkO08q6.js";
import "../../text-runtime-DVXt5xhi.js";
import "../../tool-catalog-BjSY4C4F.js";
import "../../docker-Bhjg8g2t.js";
import "../../sandbox-DTlKNieF.js";
import "../../common-zReYuoMM.js";
import "../../image-ops-By4YfyWa.js";
import "../../thinking-CJoSRIGR.js";
import "../../path-alias-guards-CwRM04O1.js";
import "../../sandbox-paths-Cn8i8Jxm.js";
import "../../channel-actions-fV93bLjV.js";
import "../../mime-CsUtzcn2.js";
import "../../ssrf-0bPJMoZR.js";
import "../../fetch-guard-CYl1q2XH.js";
import "../../provider-web-search-COZNygbd.js";
import "../../manager-Bw8JrihM.js";
import "../../config-schema-CPzP2Ds6.js";
import "../../channel-plugin-common-ClptUZHU.js";
import "../../status-helpers-Cflmo90d.js";
import "../../conversation-runtime-nmWgJSXS.js";
import "../../runtime-whatsapp-boundary-XjdkYMz_.js";
import "../../pairing-store-BGKLrjo6.js";
import "../../json-store--JDX3zNL.js";
import { t as definePluginEntry } from "../../plugin-entry-B2shVOQl.js";
import "../../secret-file-Bif7RlZ-.js";
import "../../channel-config-schema-CgpS17zB.js";
import "../../setup-binary-C2MYZTCc.js";
import "../../archive-pOxyr9r4.js";
import "../../fs-safe-DZyOgpPN.js";
import "../../signal-cli-install-fVOL9FvX.js";
import { t as resolveSecretInputModeForEnvSelection } from "../../provider-auth-mode-DkhMmaR7.js";
import "../../setup-wizard-proxy-DoOpOF8S.js";
import "../../setup-BI3ohdrT.js";
import "../../dm-policy-shared-DtcQProI.js";
import "../../runtime-env-DQOYrXVl.js";
import "../../temp-path-B91lt0LF.js";
import "../../diagnostic-CDTtMQBT.js";
import "../../logging-CCTKg_fh.js";
import "../../config-runtime-BYNizC50.js";
import "../../level-overrides-DzIzQ3ZY.js";
import "../../exec-approvals-BF_Qfdq8.js";
import "../../webhook-ingress-D-H5frou.js";
import "../../system-events-BU_NYynX.js";
import "../../ssrf-policy-DYINyIKC.js";
import "../../hook-runtime-BUYmDbaA.js";
import "../../templating-BDXRiJUh.js";
import "../../channel-reply-pipeline-BxH4-AP2.js";
import "../../reply-history-n7nOPLta.js";
import "../../commands-registry.data-BFXJWoqi.js";
import "../../commands-registry-Bh9pJ817.js";
import "../../frontmatter-CP5371Rm.js";
import "../../env-overrides-BNSo89u5.js";
import "../../skills-M0AZJeXx.js";
import "../../skills-remote-BHxDA4Jw.js";
import "../../workspace-dirs-DqDLiRvZ.js";
import "../../pairing-token-DAszGmwD.js";
import "../../skill-commands-D73dGR_l.js";
import "../../config-CLRQPUSt.js";
import "../../routes-XT2JfvT-.js";
import "../../ssh-tunnel-LM52qY6N.js";
import "../../server-middleware-BP59C6FL.js";
import { n as promptSecretRefForSetup } from "../../provider-auth-ref-CQEp9sp3.js";
import { t as applyAuthProfileConfig } from "../../provider-auth-helpers-Do3RFvdi.js";
import { t as createProviderApiKeyAuthMethod } from "../../provider-api-key-auth-BEeRW3dq.js";
import "../../pairing-labels-D27Y4Vv6.js";
import "../../directory-runtime-D9Y42mW-.js";
import "../../read-only-account-inspect-CJPRBFj7.js";
import "../../src-BoNU6eQs.js";
import "../../web-media-PKy3RbRG.js";
import "../../media-understanding-BDOSvuM9.js";
import "../../web-media-CiSbdGPs.js";
import "../../state-paths-Dvb3Mfl3.js";
import "../../llm-task-Br6Ixjlh.js";
import "../../pi-model-discovery-DO3tFksS.js";
import "../../exec-inline-eval-CpmxWJsW.js";
import "../../target-registry-9OUOvlxj.js";
import "../../external-content-DZI9vlJn.js";
import "../../brave-onFD06fq.js";
import "../../duckduckgo-CZuAOfiQ.js";
import "../../exa-5bwgY04k.js";
import "../../security-runtime-xlw0Fxx2.js";
import { a as fetchClaudeUsage } from "../../provider-usage-ClxTGn5c.js";
import "../../provider-models-Bjv0V4OP.js";
import "../../media-understanding-provider-DllGbg-v.js";
import "../../media-understanding-provider-BDq4rFL-.js";
import "../../provider-onboard-CSnPvkMM.js";
import "../../onboard-BU8aRvq7.js";
import "../../perplexity-web-search-provider-CI38RNUU.js";
import "../../perplexity-CMZbsreI.js";
import "../../model-definitions-CLagLj0g.js";
import "../../onboard-0P7GBI2D.js";
import "../../provider-catalog-CKHo9CRM.js";
import "../../provider-models-Dmjtw6kn.js";
import "../../stream-C-KGzzT0.js";
import "../../web-search-CO1hvu_l.js";
import "../../stagger-CFqdtCYL.js";
import "../../command-secret-targets-DLs8jC2H.js";
import "../../delivery-queue-DHURMSqX.js";
import "../../channel-summary-Ctrv7moa.js";
import "../../follow-redirects-CVAaqIk4.js";
import "../../tool-policy-match-DQWWRSN4.js";
import "../../runtime-CAdCdG-M.js";
import "../../channel-status-BfM90KqH.js";
import "../../discord-core-DLbIyXWy.js";
import "../../cli-runtime-Cx1v7N7v.js";
import "../../config-presence-DA3j8Nmi.js";
import "../../query-expansion-BOJBTXsp.js";
import "../../search-manager-DNBHwaee.js";
import "../../acp-runtime-BHZj-SLU.js";
import "../../telegram-core-DD8gCuaX.js";
import "../../audit-BI48Fi3j.js";
import "../../gateway-runtime-Dn2qAGY7.js";
import "../../connection-auth-Nl2a3tDb.js";
import "../../mcp-config-CDIWSqVk.js";
import { t as anthropicMediaUnderstandingProvider } from "../../media-understanding-provider-DKEqPSf8.js";
//#region extensions/anthropic/index.ts
const PROVIDER_ID = "anthropic";
const DEFAULT_ANTHROPIC_MODEL = "anthropic/claude-sonnet-4-6";
const ANTHROPIC_OPUS_46_MODEL_ID = "claude-opus-4-6";
const ANTHROPIC_OPUS_46_DOT_MODEL_ID = "claude-opus-4.6";
const ANTHROPIC_OPUS_TEMPLATE_MODEL_IDS = ["claude-opus-4-5", "claude-opus-4.5"];
const ANTHROPIC_SONNET_46_MODEL_ID = "claude-sonnet-4-6";
const ANTHROPIC_SONNET_46_DOT_MODEL_ID = "claude-sonnet-4.6";
const ANTHROPIC_SONNET_TEMPLATE_MODEL_IDS = ["claude-sonnet-4-5", "claude-sonnet-4.5"];
const ANTHROPIC_MODERN_MODEL_PREFIXES = [
	"claude-opus-4-6",
	"claude-sonnet-4-6",
	"claude-opus-4-5",
	"claude-sonnet-4-5",
	"claude-haiku-4-5"
];
const ANTHROPIC_OAUTH_ALLOWLIST = [
	"anthropic/claude-sonnet-4-6",
	"anthropic/claude-opus-4-6",
	"anthropic/claude-opus-4-5",
	"anthropic/claude-sonnet-4-5",
	"anthropic/claude-haiku-4-5"
];
function cloneFirstTemplateModel(params) {
	const trimmedModelId = params.modelId.trim();
	for (const templateId of [...new Set(params.templateIds)].filter(Boolean)) {
		const template = params.ctx.modelRegistry.find(PROVIDER_ID, templateId);
		if (!template) continue;
		return normalizeModelCompat({
			...template,
			id: trimmedModelId,
			name: trimmedModelId
		});
	}
}
function resolveAnthropic46ForwardCompatModel(params) {
	const trimmedModelId = params.ctx.modelId.trim();
	const lower = trimmedModelId.toLowerCase();
	if (!(lower === params.dashModelId || lower === params.dotModelId || lower.startsWith(`${params.dashModelId}-`) || lower.startsWith(`${params.dotModelId}-`))) return;
	const templateIds = [];
	if (lower.startsWith(params.dashModelId)) templateIds.push(lower.replace(params.dashModelId, params.dashTemplateId));
	if (lower.startsWith(params.dotModelId)) templateIds.push(lower.replace(params.dotModelId, params.dotTemplateId));
	templateIds.push(...params.fallbackTemplateIds);
	return cloneFirstTemplateModel({
		modelId: trimmedModelId,
		templateIds,
		ctx: params.ctx
	});
}
function resolveAnthropicForwardCompatModel(ctx) {
	return resolveAnthropic46ForwardCompatModel({
		ctx,
		dashModelId: ANTHROPIC_OPUS_46_MODEL_ID,
		dotModelId: ANTHROPIC_OPUS_46_DOT_MODEL_ID,
		dashTemplateId: "claude-opus-4-5",
		dotTemplateId: "claude-opus-4.5",
		fallbackTemplateIds: ANTHROPIC_OPUS_TEMPLATE_MODEL_IDS
	}) ?? resolveAnthropic46ForwardCompatModel({
		ctx,
		dashModelId: ANTHROPIC_SONNET_46_MODEL_ID,
		dotModelId: ANTHROPIC_SONNET_46_DOT_MODEL_ID,
		dashTemplateId: "claude-sonnet-4-5",
		dotTemplateId: "claude-sonnet-4.5",
		fallbackTemplateIds: ANTHROPIC_SONNET_TEMPLATE_MODEL_IDS
	});
}
function matchesAnthropicModernModel(modelId) {
	const lower = modelId.trim().toLowerCase();
	return ANTHROPIC_MODERN_MODEL_PREFIXES.some((prefix) => lower.startsWith(prefix));
}
function buildAnthropicAuthDoctorHint(params) {
	const legacyProfileId = params.profileId ?? "anthropic:default";
	const suggested = suggestOAuthProfileIdForLegacyDefault({
		cfg: params.config,
		store: params.store,
		provider: PROVIDER_ID,
		legacyProfileId
	});
	if (!suggested || suggested === legacyProfileId) return "";
	const storeOauthProfiles = listProfilesForProvider(params.store, PROVIDER_ID).filter((id) => params.store.profiles[id]?.type === "oauth").join(", ");
	const cfgMode = params.config?.auth?.profiles?.[legacyProfileId]?.mode;
	const cfgProvider = params.config?.auth?.profiles?.[legacyProfileId]?.provider;
	return [
		"Doctor hint (for GitHub issue):",
		`- provider: ${PROVIDER_ID}`,
		`- config: ${legacyProfileId}${cfgProvider || cfgMode ? ` (provider=${cfgProvider ?? "?"}, mode=${cfgMode ?? "?"})` : ""}`,
		`- auth store oauth profiles: ${storeOauthProfiles || "(none)"}`,
		`- suggested profile: ${suggested}`,
		`Fix: run "${formatCliCommand("openclaw doctor --yes")}"`
	].join("\n");
}
async function runAnthropicSetupToken(ctx) {
	await ctx.prompter.note(["Run `claude setup-token` in your terminal.", "Then paste the generated token below."].join("\n"), "Anthropic setup-token");
	const requestedSecretInputMode = normalizeSecretInputModeInput(ctx.secretInputMode);
	const selectedMode = ctx.allowSecretRefPrompt ? await resolveSecretInputModeForEnvSelection({
		prompter: ctx.prompter,
		explicitMode: requestedSecretInputMode,
		copy: {
			modeMessage: "How do you want to provide this setup token?",
			plaintextLabel: "Paste setup token now",
			plaintextHint: "Stores the token directly in the auth profile"
		}
	}) : "plaintext";
	let token = "";
	let tokenRef;
	if (selectedMode === "ref") {
		const resolved = await promptSecretRefForSetup({
			provider: "anthropic-setup-token",
			config: ctx.config,
			prompter: ctx.prompter,
			preferredEnvVar: "ANTHROPIC_SETUP_TOKEN",
			copy: {
				sourceMessage: "Where is this Anthropic setup token stored?",
				envVarPlaceholder: "ANTHROPIC_SETUP_TOKEN"
			}
		});
		token = resolved.resolvedValue.trim();
		tokenRef = resolved.ref;
	} else {
		const tokenRaw = await ctx.prompter.text({
			message: "Paste Anthropic setup-token",
			validate: (value) => validateAnthropicSetupToken(String(value ?? ""))
		});
		token = String(tokenRaw ?? "").trim();
	}
	const tokenError = validateAnthropicSetupToken(token);
	if (tokenError) throw new Error(tokenError);
	const profileNameRaw = await ctx.prompter.text({
		message: "Token name (blank = default)",
		placeholder: "default"
	});
	return { profiles: [{
		profileId: buildTokenProfileId({
			provider: PROVIDER_ID,
			name: String(profileNameRaw ?? "")
		}),
		credential: {
			type: "token",
			provider: PROVIDER_ID,
			token,
			...tokenRef ? { tokenRef } : {}
		}
	}] };
}
async function runAnthropicSetupTokenNonInteractive(ctx) {
	const provider = ctx.opts.tokenProvider?.trim().toLowerCase();
	if (!provider) {
		ctx.runtime.error("Missing --token-provider for --auth-choice token.");
		ctx.runtime.exit(1);
		return null;
	}
	if (provider !== PROVIDER_ID) {
		ctx.runtime.error("Only --token-provider anthropic is supported for --auth-choice token.");
		ctx.runtime.exit(1);
		return null;
	}
	const token = normalizeSecretInput(ctx.opts.token);
	if (!token) {
		ctx.runtime.error("Missing --token for --auth-choice token.");
		ctx.runtime.exit(1);
		return null;
	}
	const tokenError = validateAnthropicSetupToken(token);
	if (tokenError) {
		ctx.runtime.error(tokenError);
		ctx.runtime.exit(1);
		return null;
	}
	let expires;
	const expiresInRaw = ctx.opts.tokenExpiresIn?.trim();
	if (expiresInRaw) try {
		expires = Date.now() + parseDurationMs(expiresInRaw, { defaultUnit: "d" });
	} catch (err) {
		ctx.runtime.error(`Invalid --token-expires-in: ${String(err)}`);
		ctx.runtime.exit(1);
		return null;
	}
	const profileId = ctx.opts.tokenProfileId?.trim() || buildTokenProfileId({
		provider: PROVIDER_ID,
		name: ""
	});
	upsertAuthProfile({
		profileId,
		agentDir: ctx.agentDir,
		credential: {
			type: "token",
			provider: PROVIDER_ID,
			token,
			...expires ? { expires } : {}
		}
	});
	return applyAuthProfileConfig(ctx.config, {
		profileId,
		provider: PROVIDER_ID,
		mode: "token"
	});
}
var anthropic_default = definePluginEntry({
	id: PROVIDER_ID,
	name: "Anthropic Provider",
	description: "Bundled Anthropic provider plugin",
	register(api) {
		api.registerProvider({
			id: PROVIDER_ID,
			label: "Anthropic",
			docsPath: "/providers/models",
			envVars: ["ANTHROPIC_OAUTH_TOKEN", "ANTHROPIC_API_KEY"],
			deprecatedProfileIds: [CLAUDE_CLI_PROFILE_ID],
			auth: [{
				id: "setup-token",
				label: "setup-token (claude)",
				hint: "Paste a setup-token from `claude setup-token`",
				kind: "token",
				wizard: {
					choiceId: "token",
					choiceLabel: "Anthropic token (paste setup-token)",
					choiceHint: "Run `claude setup-token` elsewhere, then paste the token here",
					groupId: "anthropic",
					groupLabel: "Anthropic",
					groupHint: "setup-token + API key",
					modelAllowlist: {
						allowedKeys: [...ANTHROPIC_OAUTH_ALLOWLIST],
						initialSelections: ["anthropic/claude-sonnet-4-6"],
						message: "Anthropic OAuth models"
					}
				},
				run: async (ctx) => await runAnthropicSetupToken(ctx),
				runNonInteractive: async (ctx) => await runAnthropicSetupTokenNonInteractive({
					config: ctx.config,
					opts: ctx.opts,
					runtime: ctx.runtime,
					agentDir: ctx.agentDir
				})
			}, createProviderApiKeyAuthMethod({
				providerId: PROVIDER_ID,
				methodId: "api-key",
				label: "Anthropic API key",
				hint: "Direct Anthropic API key",
				optionKey: "anthropicApiKey",
				flagName: "--anthropic-api-key",
				envVar: "ANTHROPIC_API_KEY",
				promptMessage: "Enter Anthropic API key",
				defaultModel: DEFAULT_ANTHROPIC_MODEL,
				expectedProviders: ["anthropic"],
				wizard: {
					choiceId: "apiKey",
					choiceLabel: "Anthropic API key",
					groupId: "anthropic",
					groupLabel: "Anthropic",
					groupHint: "setup-token + API key"
				}
			})],
			resolveDynamicModel: (ctx) => resolveAnthropicForwardCompatModel(ctx),
			capabilities: {
				providerFamily: "anthropic",
				dropThinkingBlockModelHints: ["claude"]
			},
			isModernModelRef: ({ modelId }) => matchesAnthropicModernModel(modelId),
			resolveDefaultThinkingLevel: ({ modelId }) => matchesAnthropicModernModel(modelId) && (modelId.toLowerCase().startsWith(ANTHROPIC_OPUS_46_MODEL_ID) || modelId.toLowerCase().startsWith(ANTHROPIC_OPUS_46_DOT_MODEL_ID) || modelId.toLowerCase().startsWith(ANTHROPIC_SONNET_46_MODEL_ID) || modelId.toLowerCase().startsWith(ANTHROPIC_SONNET_46_DOT_MODEL_ID)) ? "adaptive" : void 0,
			resolveUsageAuth: async (ctx) => await ctx.resolveOAuthToken(),
			fetchUsageSnapshot: async (ctx) => await fetchClaudeUsage(ctx.token, ctx.timeoutMs, ctx.fetchFn),
			isCacheTtlEligible: () => true,
			buildAuthDoctorHint: (ctx) => buildAnthropicAuthDoctorHint({
				config: ctx.config,
				store: ctx.store,
				profileId: ctx.profileId
			})
		});
		api.registerMediaUnderstandingProvider(anthropicMediaUnderstandingProvider);
	}
});
//#endregion
export { anthropic_default as default };
